frontend/src/settings.rs

```diff
@@ -1,107 +1,291 @@
+pub enum KeyboardMode {
+    Standard,
+    Vim,
+    Emacs,
+}
+
+impl DisplayLocalized for KeyboardMode {
+    fn to_localized_string(&self, locale: Locale) -> String {
+        match self {
+            KeyboardMode::Vim => td_display!(locale, vim_mode),
+            KeyboardMode::Emacs => td_display!(locale, emacs_mode),
+            KeyboardMode::Standard => td_display!(locale, standard_mode),
+        }
+        .to_string()
     }
 }
 
-#[component]
-fn LocaleSelector() -> impl IntoView {
-    fn locale_name(locale: Locale) -> &'static str {
-        match locale {
-            Locale::en => "English",
-            Locale::it => "Italiano",
-            Locale::es => "Español",
-            Locale::ca => "Català",
-            Locale::vec => "Vèneto",
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Debug, Serialize, Deserialize, VariantArray)]
+pub enum InputMode {
+    Batch,
+    MixedInteractive,
+    FullInteractive,
+}
+
+impl DisplayLocalized for InputMode {
+    fn to_localized_string(&self, locale: Locale) -> String {
+        match *self {
+            InputMode::Batch => td_display!(locale, batch_input),
+            InputMode::MixedInteractive => td_display!(locale, mixed_interactive_input),
+            InputMode::FullInteractive => td_display!(locale, full_interactive_input),
         }
+        .to_string()
     }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+pub struct Settings {
+    theme: Option<Theme>,
+    keyboard_mode: KeyboardMode,
+    input_mode: InputMode,
+    locale: Locale,
+    editor_width_percent: f32,
+    language: Language,
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Self {
+            theme: None,
+            keyboard_mode: KeyboardMode::Standard,
+            input_mode: InputMode::Batch,
+            locale: Locale::en,
+            editor_width_percent: 65.0,
+            language: Language::CPP,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct SettingsProvider {
+    write: WriteSignal<Settings>,
+    read: Signal<Settings>,
+    pub theme: Signal<Theme>,
+    pub editor_width_percent: Signal<f32>,
+    pub keyboard_mode: Signal<KeyboardMode>,
+    pub input_mode: Signal<InputMode>,
+    pub language: Signal<Language>,
+}
+
+impl SettingsProvider {
+    pub fn install() {
+        let (read_settings, write_settings, _) =
+            use_local_storage::<Settings, JsonSerdeCodec>("wasm_ide_settings");
+        let prefers_dark = use_preferred_dark();
+        let theme = Memo::new(move |_| {
+            read_settings.get().theme.unwrap_or(if prefers_dark.get() {
+                Theme::Dark
+            } else {
+                Theme::Light
+            })
+        });
+        // Sync the theme to the entire page
+        Effect::new(move || {
+            info!(theme = ?theme.get());
+            web_sys::window()
+                .unwrap()
+                .document()
+                .unwrap()
```

Comment:
```suggestion
           document()
```
See https://docs.rs/leptos/0.8.12/leptos/prelude/fn.document.html